refactor(app,components,labware-library): Support labware-schema-3-style positioning in some components

SyntaxColoring opened the pull request:
## Overview

Goes towards EXEC-1278.

## Details

To prepare for labware schema 3, we want to move math and logic out of React components and into new utils in `shared-data`. Those utils can then abstract over differences between labware schemas 2 and 3, and implement whatever positioning features we add to labware schema 3.

By "math and logic," I mean things like: What do you choose as the SVG `viewbox`? Within that `viewbox`, how do you position the SVG elements comprising the labware? If you need to render a labware in a slot, on a module, or stacked atop another labware, how does that affect things?

There are basically two cases. Each case already has a util in `shared-data`, added in #18590.

1. We're rendering a labware on its own, like in the thumbnails in the labware gallery.
2. We're rendering a labware relative to other things, like on a deck map.

So, in this PR:

* We switch a bunch of components from manually constructing `viewBox`es and `transform`s to delegating to the appropriate util from `shared-data`.
* We set `<Labware positioningMode="passThrough" ... />`, which the new utils assume. See EXEC-1573 for background.

This is the first of probably several PRs that will do this. The selection of components so far is basically random. 

## Test Plan and Hands on Testing

Spot-check individual components and make sure their current real-world usage, rendering labware schema 2, is unaffected. Ideally, we check labware schema 3 too.

* [x] app's `<LabwareCard>` and `<Gallery>`
    * [x] Various labware's list entries in the Labware tab
    * [x] And those labwares' details slideout
    * [x] And a schema 3 definition
* [x] labware-library's `<Gallery>`
* [x] `<CalibrationBlockRender>` and `<CalibrationLabwareRender>`
    * [x] Storybook
    * [x] In-context in OT-2 calibration flows
* [x] `<WellSelection>`
    * [x] In-context, in Quick Transfer
* [x] `<LabwareListItem>`
* [x] `<SlotDetailModal>`
* [x] `<LabwareLiquidsDetailModal>`


## Review requests

* Does the new pattern make sense?
* Anything special about the components modified so far that we have to watch out for?

## Risk assessment

Medium.

SyntaxColoring committed:
Explain opentrons_universal_flat_adapter xDimension override.

SyntaxColoring committed:
Document RobotWorkSpace.

SyntaxColoring committed:
Update app's <LabwareCard> and <Gallery>.

SyntaxColoring committed:
Audit all uses of Labware and LabwareRender and leave todo comments.

SyntaxColoring committed:
Refactor a hard-coded deck viewbox into a constant.

SyntaxColoring committed:
vectorMath: Add coordinateTupleToVector3D() converter.

SyntaxColoring committed:
Correct <RobotWorkSpace>'s deckSlotsById to addressableAreasById.

This just changes the name and type to match what the object actually is.
e.g. this includes things like module addressable areas, which are not deck slots.

SyntaxColoring committed:
Document RobotCoordsText.

I confirmed that deleted todo comment got resolved a while ago.

SyntaxColoring committed:
Update <CalibrationBlockRender>.

SyntaxColoring committed:
Update <SlotDetailModal>.

SyntaxColoring committed:
Update <LabwareLiquidsDetailModal>.

SyntaxColoring committed:
Clarify LabwareRender docs.

SyntaxColoring committed:
Update labware-library's <Gallery>.

SyntaxColoring committed:
Update <CalibrationLabwareRender>.

SyntaxColoring committed:
Update <WellSelection>.

SyntaxColoring committed:
Update <LabwareListItem>.

SyntaxColoring committed:
Delete remaining todo comments. Deferring to separate PRs.

mjhuff commented on app/src/organisms/Desktop/CalibrationPanels/DeckSetup.tsx:
```diff
@@ -109,31 +112,56 @@ export function DeckSetup(props: CalibrationPanelProps): JSX.Element {
+                  }
+                  if (labwareDef === null) {
+                    return null
+                  }
```
Would much prefer this to be an IIFE or a full function definition with invocation than to use `let` bindings.

SyntaxColoring replied:
61f933e4ec6b3c240971140634e9148839549461, let me know if that's what you were going for.

mjhuff replied:
Perf 👌

mjhuff commented on app/src/organisms/Desktop/CalibrationPanels/DeckSetup.tsx:
```diff
@@ -109,31 +112,56 @@ export function DeckSetup(props: CalibrationPanelProps): JSX.Element {
+                    addressableArea.id,
+                    deckDef
+                  )
+                  if (slotOrigin === null) {
```
```suggestion
                  if (slotOrigin == null) {
```

mjhuff commented on app/src/organisms/Desktop/Labware/LabwareDetails/Gallery.tsx:
```diff
@@ -11,15 +11,18 @@ import {
+ * replacement xDimension that includes the protrusion so it doesn't get clipped off
+ * when we render an SVG of the adapter.
+ */
 export const UNIVERSAL_FLAT_ADAPTER_X_DIMENSION = 127.4
```
Thanks for the docstring here.

mjhuff commented on components/src/hardware-sim/Deck/RobotWorkSpace.tsx:
```diff
@@ -23,6 +34,14 @@ export interface RobotWorkSpaceProps extends StyleProps {
+ * SVG-transform them so they're displayed the correct way. This is needed because
+ * SVG inverts y compared to Opentrons coordinates.
+ */
 export function RobotWorkSpace(props: RobotWorkSpaceProps): JSX.Element | null {
```
Wasn't a part of your PR, but could we get rid of the ` | null` here?

SyntaxColoring replied:
https://github.com/Opentrons/opentrons/pull/18557/commits/5619fb679b68f113025505d5a1b460d3ae47b9fd

This took a little bit of finagling to convince TypeScript that we would always have at least one of `viewBox` or `deckDef` to fall back on, making the `if (!deckDef && !viewBox) return null` case unreachable.

It relies on the `in` operator, which I'm a little uncomfortable with because it seems technically unsound? (TypeScript narrows `"viewBox" in props === true` into `props: {viewBox: string}`, which seems nice, except that I think there are ways to sneak in a `viewBox` value isn't a string, so that narrowing isn't actually always true.) Let me know if there's a better pattern for this kind of thing.

mjhuff replied:
Yeah that is tricky. What about doing something like:

```
function assertHasViewBox(
  props: RobotWorkSpaceProps
): asserts props is RobotWorkSpaceProps & { viewBox: string } {
  if (!('viewBox' in props) || typeof props.viewBox !== 'string') {
    throw new Error('Expected viewBox to be present and be a string')
  }
}
```

or perhaps less aggressively:
```
function hasViewBox(props: RobotWorkSpaceProps): props is RobotWorkSpaceProps & { viewBox: string } {
  return 'viewBox' in props && typeof props.viewBox === 'string'
}
```

SyntaxColoring replied:
After meditating on this, I'm going to keep it as in 5619fb679b68f113025505d5a1b460d3ae47b9fd and not worry too hard about the unsoundness. TypeScript is unsound sometimes and that makes me sad, but I guess it's not a problem in practice (even though there are [very simple ways to get bitten by it](https://www.typescriptlang.org/play/?#code/FAYw9gdgzgLgBAQzgXjgbwGYEYBccBESsATgJYQDm+ANHBgEx5b0DMAvsKJLHAEZ6ZccEuQpsUiLtHggB2PCMq0GAfgUwylcal6dSGOAAoQAOgZwAhMlQBXCABMAphnKP7ASnTA4cU+agIMKRQLo5QwhqicAD00XAA7sSQFMr0cMGIIDA2CAA2uQCeiHAQNgC2vI7EcDiGwBxAA)), so it doesn't make sense to fight unless we have a specific reason to. 🤷

mjhuff commented on components/src/hardware-sim/Deck/RobotWorkSpace.tsx:
```diff
@@ -38,12 +57,12 @@ export function RobotWorkSpace(props: RobotWorkSpaceProps): JSX.Element | null {
 
   let wholeDeckViewBox
-  let deckSlotsById = {}
+  let addressableAreasById = {}
```
Could we get rid of the `let` binding here entirely?

mjhuff commented on components/src/hardware-sim/Deck/RobotWorkSpace.tsx:
```diff
@@ -6,15 +6,26 @@ import { Svg } from '../../primitives'
 
 export interface RobotWorkSpaceRenderProps {
-  deckSlotsById: { [slotId: string]: DeckSlot }
+  // todo(mm, 2025-06-05): Is this API still worthwhile? The parent of RobotWorkSpace
```
Yeah, I agree with you here. Doesn't have to be in this PR, but this iface does seem unnecessary now.

mjhuff reviewed (approved):
Nice, this all seems reasonable. Agree with the overall approach and test plan. Just a few clean up comments.

SyntaxColoring committed:
Refactor `let` ladder into an IIFE. Also simplify the logic inside it.

I'm pretty sure this was sometimes comparing the string "undefined" against addressable area name strings.

SyntaxColoring committed:
Compare against null with `==` instead of `===`.

SyntaxColoring committed:
Clean up RobotWorkSpace.

- Type props more precisely to remove the possible `null` return.
- Clean up internal logic to avoid `let` bindings.

SyntaxColoring committed:
Delete broken mocks?

SyntaxColoring committed:
Delete more broken mocks?

SyntaxColoring commented on app/src/organisms/Desktop/CalibrateDeck/__tests__/CalibrateDeck.test.tsx:
```diff
@@ -91,7 +82,6 @@ describe('CalibrateDeck', () => {
 
   beforeEach(() => {
     dispatchRequests = vi.fn()
-    vi.mocked(getDeckDefinitions).mockReturnValue({})
```
These tests were rendering something with `<RobotWorkSpace>` in the component tree. Something in the component tree was calling `getDeckDefinitions()`, finding an OT-2 definition, and passing it to `<RobotWorkSpace>`.

These mocks were breaking that because, of course, `.mockReturnValue({})` doesn't contain an OT-2 deck definition. By luck, we were getting away with this until 5619fb679b68f113025505d5a1b460d3ae47b9fd.

My solution is to just...delete the mock? I don't know why these tests wanted to mock out `getDeckDefinitions()` in the first place, and I don't know how we would want to fix the mocks even if we wanted to keep them.

SyntaxColoring commented on app/src/organisms/Desktop/CalibrationPanels/__tests__/DeckSetup.test.tsx:
```diff
@@ -13,17 +13,6 @@ import {
 import type { ComponentProps } from 'react'
-import type { getDeckDefinitions } from '@opentrons/shared-data'
-
-vi.mock('/app/assets/labware/getLabware')
-vi.mock('@opentrons/shared-data', async importOriginal => {
-  const actual = await importOriginal<typeof getDeckDefinitions>()
-  return {
-    ...actual,
-    getDeckDefinitions: () => vi.fn(),
-  }
-})
-vi.mock('@opentrons/components/src/hardware-sim/Deck/RobotWorkSpace')
```
A similar deal here.

* The `@opentrons/components/src/hardware-sim/Deck/RobotWorkSpace` mock was never doing anything because that's not how `RobotWorkSpace` is imported. It would have needed to be a partial mock of `@opentrons/components`, I think.
* The `getDeckDefinitions()` and `getLabware()` mocks were not configured anywhere in this file.
